nodeletctl/pkg/nodeletctl/nodeletctl.go

```diff
@@ -106,6 +106,130 @@ func CreateCluster(cfgPath string) error {
+		if host.NodeIP != nil {
+			masterList[host.NodeName] = *host.NodeIP
+		} else {
+			masterList[host.NodeName] = host.NodeName
+		}
+	}
+
+	masters, err := GetCurrentMasters(clusterCfg)
+	if err != nil {
+		return fmt.Errorf("Failed to get active K8s masters: %s", err)
+	}
+
+	workers, err := GetCurrentWorkers(clusterCfg)
+	if err != nil {
+		return fmt.Errorf("Failed to get active K8s workers: %s", err)
+	}
+
+	if len(clusterCfg.MasterNodes) != len(masters) {
+		return fmt.Errorf("Masters in the cluster and Bootstrap config doesnt match", err)
+	}
+
+	if len(clusterCfg.WorkerNodes) != len(workers) {
+		return fmt.Errorf("Workers in the cluster and Bootstrap config doesnt match", err)
+	}
+
+	// Upgrade each master node sequentially
+	for _, host := range clusterCfg.MasterNodes {
+		zap.S().Infof("Upgrading master node: ", host.NodeName)
+		// Build nodeletconfig and deployer
+		nodeletCfg := new(NodeletConfig)
+		setNodeletClusterCfg(clusterCfg, nodeletCfg)
+		nodeletCfg.HostId = host.NodeName
+		if host.NodeIP != nil {
+			nodeletCfg.HostIp = *host.NodeIP
+		} else {
+			nodeletCfg.HostIp = host.NodeName
+		}
+		nodeletCfg.NodeletRole = "master"
+		nodeletCfg.MasterList = &masterList
+		nodeletCfg.EtcdClusterState = "new"
+
+		nodeletSrcFile, err := GenNodeletConfigLocal(nodeletCfg, masterNodeletConfigTmpl)
+		if err != nil {
+			zap.S().Infof("Failed to generate config: %s", err)
+			return fmt.Errorf("Failed to generate config: %s", err)
+		}
+		zap.S().Debugf("master nodeletsrc file %s", nodeletSrcFile)
+
+		deployer, err := GetNodeletDeployer(clusterCfg, globalClusterStatus, nodeletCfg, nodeletCfg.HostIp, nodeletSrcFile)
+		if err != nil {
+			zap.S().Errorf("failed to get nodelet deployer: %v", err)
+			return fmt.Errorf("failed to get nodelet deployer: %v", err)
+		}
+		globalClusterStatus.statusMap[host.NodeName] = &NodeStatus{
+			deployer: deployer,
+		}
+
+		if err := deployer.UpgradeNode(); err != nil {
+			SetClusterNodeStatus(deployer.clusterStatus, deployer.nodeletCfg.HostId, "failed", err)
+		}
+	}
+
+	// Upgrade each worker node sequentially
+	for _, host := range clusterCfg.WorkerNodes {
+		zap.S().Infof("Upgrading worker: ", host.NodeName)
+		nodeletCfg := new(NodeletConfig)
+		setNodeletClusterCfg(clusterCfg, nodeletCfg)
+		nodeletCfg.HostId = host.NodeName
+		nodeletCfg.NodeletRole = "worker"
+		nodeletSrcFile, err := GenNodeletConfigLocal(nodeletCfg, workerNodeletConfigTmpl)
+		if err != nil {
+			zap.S().Infof("Failed to generate config: %s", err)
+			return fmt.Errorf("Failed to generate config: %s", err)
+		}
+		deployer, err := GetNodeletDeployer(clusterCfg, globalClusterStatus, nodeletCfg, host.NodeName, nodeletSrcFile)
+		if err != nil {
+			zap.S().Errorf("failed to get nodelet deployer: %s", err)
+			return fmt.Errorf("failed to get nodelet deployer: %s", err)
+		}
+		globalClusterStatus.statusMap[host.NodeName] = &NodeStatus{
+			deployer: deployer,
+		}
+
+		if err := deployer.UpgradeNode(); err != nil {
```

Comment:
try to make this in parallel like when we create or scale, or do in batches. I don't think we need to do workers one by one unless theres a specific reason